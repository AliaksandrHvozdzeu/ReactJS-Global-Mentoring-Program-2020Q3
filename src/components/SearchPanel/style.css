.search {
    margin: 0 5%;
    height: 80%;
    padding: 2% 0;
}

.search-field {
    height: 3rem;
    width: 84%;
    color: var(--white);
    background-color: var(--grey-darker);
    font-size: 1.2rem;
    border-radius: 4px;
    box-shadow: 0 0 10px 0 #2323239a;
    opacity: 0.8;
    outline: none;
    padding: 0 0.5rem;
    margin-bottom: 1rem;
}

.search-panel {
    margin-top: 2rem;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    flex-wrap: wrap;
}

.search-title {
    display: block;
    color: var(--white);
    text-transform: uppercase;
    font-weight: lighter;
    font-size: 2.5rem;
    letter-spacing: 0.1rem;
}
