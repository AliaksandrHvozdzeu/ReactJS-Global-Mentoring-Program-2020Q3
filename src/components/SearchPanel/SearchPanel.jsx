import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import AddButton from '../AddButton';
import Brand from '../Brand';
import Wrapper from '../Wrapper';
import './style.css';

const SearchPanel = ({ filterByName, onFilterByName }) => {

  const [searchString, setSearchString] = useState(filterByName);

  const handleKeyDown = (event) => {
    if (event.key === 'Enter') {
      onFilterByName(searchString);
    }
  };

  const onSearch = useCallback(() => onFilterByName(searchString), [
    searchString,
    onFilterByName,
  ]);

  return (
    <>
      <Wrapper>
        <Brand/>
        <div className="search">
            <label htmlFor="search-field" className="search-title">
              FIND YOUR MOVIE
              <div className="search-panel">
                <input type="text"
                       id="search-field"
                       className="search-field"
                       placeholder="What do you want to watch?"
                       onKeyDown={handleKeyDown}
                       value={searchString}
                       onChange={(e) => setSearchString(e.target.value)}/>
                <button type="button" className="search-button" onClick={onSearch}>Search</button>
                <AddButton/>
              </div>
            </label>
        </div>
      </Wrapper>
    </>
  );
};

SearchPanel.propTypes = {
  filterByName: PropTypes.string.isRequired,
  onFilterByName: PropTypes.func.isRequired,
};

export default SearchPanel;
