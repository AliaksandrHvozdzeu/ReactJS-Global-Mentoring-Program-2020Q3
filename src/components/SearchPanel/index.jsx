import React from 'react';
import AddButton from '../Buttons/AddButton';
import SearchButton from '../Buttons/SearchButton';
import SearchInput from '../Inputs/SearchInput';
import './style.css';

export default () => (
  <div className="search">
    <label htmlFor="search-field" className="search-title">
      FIND YOUR MOVIE
      <div className="search-panel">
        <SearchInput/>
        <SearchButton/>
        <AddButton/>
      </div>
    </label>
  </div>
);
