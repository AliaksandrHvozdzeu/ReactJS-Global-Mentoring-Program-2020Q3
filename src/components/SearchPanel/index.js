import SearchPanel from './js/SearchPanel';

export default SearchPanel;
