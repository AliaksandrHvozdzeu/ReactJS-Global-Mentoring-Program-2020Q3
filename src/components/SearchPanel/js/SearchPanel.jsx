import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AddButton from '../../AddButton';
import Brand from '../../Brand';
import Wrapper from '../../Wrapper';
import { movieActions } from '../../../store/actions';
import '../css/SearchPanel.css';

let searchValue = null;

const SearchPanel = ({ onSearch, loadMoviesAfterClearSearch }) => {

  const [searchString, setSearchString] = useState('');
  const [isMessageFrame, setIsMessageFrame] = useState(false);

  const handleKeyDown = useCallback(
    (event) => {
      if (event.key === 'Enter') {
        onSearch();
      }
    },
    [searchString],
  );

  const showMessage = () => {
    setIsMessageFrame(true);
  };

  const handleSearch = () => {
    return searchValue === '' || searchValue === null || searchValue === undefined ? (
      <button type="button" className="search-button" onClick={showMessage}>Search</button>
    ) : (
      <button type="button" className="search-button" onClick={onSearch}>Search</button>
    );
  };

  const onClearSearchField = (target) => {
    if (target.value === '') {
      loadMoviesAfterClearSearch();
    }
  };

  return (
    <>
      <Wrapper>
        <Brand/>
        <div className="search">
          <label htmlFor="search-field" className="search-title">
            FIND YOUR MOVIE
            <div className="search-panel">
              <input id="search-field"
                     type="search"
                     className="search-field"
                     placeholder="What do you want to watch?"
                     onKeyDown={handleKeyDown}
                     value={searchString}
                     onInput={(e) => {
                       onClearSearchField(e.target);
                     }}
                     onChange={(e) => {
                       searchValue = e.target.value;
                       setSearchString(e.target.value);
                       setIsMessageFrame(false);
                     }}/>
              {handleSearch()}
              <AddButton/>
              {isMessageFrame && (
                <div role="presentation"
                     className="search-input-message"
                     onClick={() => setIsMessageFrame(false)}>
                  Please enter the part of title or full title of the movie to search!
                </div>
              )}
            </div>
          </label>
        </div>
      </Wrapper>
    </>
  );

};

SearchPanel.propTypes = {
  onSearch: PropTypes.func.isRequired,
  loadMoviesAfterClearSearch: PropTypes.func.isRequired,
};

SearchPanel.defaultProps = {};

const mapStateToProps = (state) => ({
  movies: state.searchString,
});

const mapDispatchToProps = (dispatch) => ({
  onSearch: () => dispatch(movieActions.getMovieByTitle(searchValue)),
  loadMoviesDefault: () => dispatch(movieActions.loadMoviesDefault()),
  loadMoviesAfterClearSearch: () => dispatch(movieActions.loadMoviesAfterClearSearch()),
});


export default connect(mapStateToProps, mapDispatchToProps)(SearchPanel);
