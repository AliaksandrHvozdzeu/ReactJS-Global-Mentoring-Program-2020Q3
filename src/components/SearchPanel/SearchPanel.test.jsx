import React from "react";
import SearchPanel from "./js/SearchPanel";
import { jest } from "@jest/globals";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import rerender from "react-test-renderer";
import { mount } from "enzyme";

describe("when SearchPanel", () => {

  const initialStore = {
    movies: {
      "genres": ["Comedy", "Crime", "Documentary", "Horror"]
    }
  };
  const onSearch = jest.fn();
  const onClearSearch = jest.fn();
  const store = configureStore([])(initialStore);
  store.dispatch = jest.fn();

  test("then snapshot created", () => {
    const component = rerender.create(
      <Provider store={store}>
        <SearchPanel onSearch={onSearch}
                     onClearSearch={onClearSearch} />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Search input onChange event", () => {
    const component = mount(<Provider store={store}>
      <SearchPanel onSearch={onSearch} onClearSearch={onClearSearch} />
    </Provider>);
    const input = component.find("#search-field");
    const e = {
      preventDefault() {},
      target: { value: "jedi" }
    };
    input.simulate("change", e.target);
    input.update();
    expect(input.text()).toEqual("");
  });

  test("Search input onInput event", () => {
    const component = mount(<Provider store={store}>
      <SearchPanel onSearch={onSearch} onClearSearch={onClearSearch} />
    </Provider>);
    const input = component.find("#search-field");
    const e = {
      preventDefault() {},
      target: { value: "jedi" }
    };
    input.simulate("input", e.target);
    input.update();
    expect(input.text()).toEqual("");
  });

});
