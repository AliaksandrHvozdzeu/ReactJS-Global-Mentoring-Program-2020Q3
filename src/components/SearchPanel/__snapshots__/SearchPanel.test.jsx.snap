// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when SearchPanel then snapshot created 1`] = `
<div
  className="wrapper unselectable"
>
  <div
    className="brand"
  >
    <p
      className="brand-title"
    >
      <a
        className="brand-title-name"
        href=" #"
        id="blink"
        onClick={[Function]}
      >
        netflix
      </a>
      roulette
    </p>
  </div>
  <div
    className="search"
  >
    <label
      className="search-title"
      htmlFor="search-field"
    >
      FIND YOUR MOVIE
      <div
        className="search-panel"
      >
        <input
          className="search-field"
          id="search-field"
          onChange={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          placeholder="What do you want to watch?"
          type="search"
          value=""
        />
        <button
          className="search-button"
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
        <button
          className="add-button"
          onClick={[Function]}
          type="button"
        >
          + ADD MOVIE
        </button>
      </div>
    </label>
  </div>
</div>
`;
